htdocs/core/tpl/contacts.tpl.php

```diff
@@ -181,7 +183,18 @@
 		<div class="tagtd nowrap noborderbottom">
 			<?php
 			$selectedCompany = GETPOSTISSET("newcompany") ? GETPOSTINT("newcompany") : (empty($object->socid) ? 0 : $object->socid);
-			$selectedCompany = $formcompany->selectCompaniesForNewContact($object, 'id', $selectedCompany, 'newcompany', array(), 0, '', 'minwidth300imp maxwidth400 widthcentpercentminusx');	// This also print the select component?>
+			$allowed_thirdparties = [];
+			if ($object->restrictiononfksoc && ! $permission_view_all_thirdparties) {
+				$allowed_thirdparties = $user->getAffectedThirdparties(1);
+				if (is_int($allowed_thirdparties)) {
+					print $user->errorsToString();
+					$allowed_thirdparties = [$object->socid];
+				} elseif (empty($allowed_thirdparties)) {
+					$allowed_thirdparties = [$object->socid];
+				}
+			}
+			$selectedCompany = $formcompany->selectCompaniesForNewContact($object, 'id', $selectedCompany, 'newcompany', $allowed_thirdparties, 0, '', 'minwidth300imp maxwidth400 widthcentpercentminusx');	// This also print the select component
```

Comment:
The test should be included into the selectCompaniesForNewContact for 2 reasons:
- it is a security condition, so should never be bypassed so should not be decided by each caller of selectCompaniesForNewContact but by selectCompaniesForNewContact itself.
- second reason is that getAffectedThirdparties can return millions of lines making memory allocation for million of php object. Will generate a memory error on large database. Filter must be directly inside the SELECT inside the selectCompaniesForNewContact.

Reply:
Thanke. I have implemented things your way.